### PR TITLE
feat: Img backend

### PR DESCRIPTION
**Pull request by andy-t-wang**
Backend updates for image upload workflow. Front end needs clean up but is currently a draft pr in img-upload-dev if you want to see how that connects. #458 (most of it is implemented just needs some clean up and testing)

Also this currently accepts PNG only since we need to make a separate post policy if we want to exactly enforce content type. I figure this is fine since PNG accommodates transparent bg, but let me know if you want to also accept jpeg. 

This pr ended up being quite large. 
1. Migrated to hasura custom actions
2. Update Precheck
3. New Tests

Might be worth holding off on adding delete images until we have permissions merged
complete WID-658
Merge #463 first

### Endpoints Overview
1. delete_images: Used to delete all unverified images when the user deletes the app
4. get_images: Gets a single image that the user specifies. Used after the upload so the user can verify the image was uploaded correctly
5. get_unverified: Checks the unverified images available on the app using the app_id and returns them. Used when the user loads an app with existing images
6. upload_image: Used when the user uploads an image.

### CODE REVIEW

File: hasura/metadata/databases/default/tables/public_app_metadata.yaml
Comment (andy-t-wang) [inline hasura/metadata/databases/default/tables/public_app_metadata.yaml:1]: Service needs to fetch unverified app images

File: web/next.config.js
Comment (pdtfh) [inline web/next.config.js:7]: S3 Bucket URL should be constructed from bucket name, `ASSETS_S3_BUCKET_NAME`

File: web/next.config.js
Comment (pdtfh) [inline web/next.config.js:8]:
```suggestion
const cdnUrl = process.env.ASSETS_CDN_URL;
```

File: web/next.config.js
Comment (pdtfh) [inline web/next.config.js:47]: let's add a note that `world-id-public.s3.amazonaws.com` is deprecated and should be removed soon

File: web/next.config.js
Comment (pdtfh) [inline web/next.config.js:21]: review with security team to understand any potential concerns

File: web/src/api/images/get_uploaded_image.ts
Comment (pdtfh) [inline web/src/api/images/get_uploaded_image.ts:45]: couldn't you pass directly req.query here and avoid the additional destructure above?

File: web/src/api/images/get_uploaded_image.ts
Comment (pdtfh) [inline web/src/api/images/get_uploaded_image.ts:53]: I don't think this is compatible with #453, we'll be pulling the team_id from the URL path, and API routes from a header, correct @igorosip0v ?

File: web/src/api/images/get_uploaded_image.ts
Comment (pdtfh) [inline web/src/api/images/get_uploaded_image.ts:63]: if we're using these errors, then this API routes should be exposed through Hasura custom actions (which I think may be a good idea) instead of as direct API routes. thoughts @igorosip0v ?

File: web/src/api/images/get_uploaded_image.ts
Comment (pdtfh) [inline web/src/api/images/get_uploaded_image.ts:56]: reminder for @pdtfh, re-review permissions conditions after rest of feedback is addressed

File: web/src/api/images/get_uploaded_image.ts
Comment (pdtfh) [inline web/src/api/images/get_uploaded_image.ts:87]: rationale? this should be short use right? maybe 15 mins?

File: web/src/api/images/get_uploaded_image.ts
Comment (pdtfh) [inline web/src/api/images/get_uploaded_image.ts:92]:
```suggestion
        success: true,
```
following convention on other endpoints

File: web/src/api/images/get_uploaded_image.ts
Comment (andy-t-wang) [reply web/src/api/images/get_uploaded_image.ts:45]: I put req.query there but still destructured it after since I use both the values in the code later. Probably slightly cleaner now

File: web/src/api/images/get_uploaded_image.ts
Comment (igorosip0v) [reply web/src/api/images/get_uploaded_image.ts:58]: Yes, but while we don't have these changes in this PR we can leave as it is. Or one thing that we can do here is accept team id from header/body/param and pass `session?.user.hasura.team_id` for now, but change to `router.query.team_id` later

File: web/src/api/images/get_uploaded_image.ts
Comment (igorosip0v) [reply web/src/api/images/get_uploaded_image.ts:68]: Mostly it will be just a wrapper, but this will give us the possibility to generate hooks to gql query and then use it on the frontend side, so we can do this.

Comment (pdtfh) [reply web/src/api/images/get_uploaded_image.ts:67]: as aligned on stand up, let's convert these endpoints to GraphQL custom actions

File: web/src/api/images/delete_all_images.ts
Comment (pdtfh) [inline web/src/api/images/delete_all_images.ts:29]: 1. how does a user then delete a single image? 
2. this should be an internal call, triggered from a delete hook from Hasura. also just in case, might be best to expire the images and let's add a lifecycle rule to the bucket (infra) to delete all expired objects after 2-3 days.

File: web/src/api/images/upload_image.ts
Comment (pdtfh) [inline web/src/api/images/upload_image.ts:40]: FYI this would normally require CORS but given you're changing it to Hasura custom actions we should be fine

File: web/src/api/images/upload_image.ts
Comment (pdtfh) [inline web/src/api/images/upload_image.ts:86]: worth adding support for jpg. if images don't have background generally this is a more effective format

**Review by pdtfh — CHANGES_REQUESTED** [review]
good progress, main change is converting to custom GraphQL actions, comments inline

File: web/src/api/images/delete_all_images.ts
Comment (andy-t-wang) [reply web/src/api/images/delete_all_images.ts:29]: I was mainly using this to clean up old resources if the user deleted the app. For verified images they will overwrite the old image to replace it and can't just delete an image. For unverified, the bucket path is protected and the hasura is updated to not have an img_url

Comment (andy-t-wang) [reply web/src/api/images/delete_all_images.ts:29]: Figure not clearing the image would accommodate for the use case where they click the x on the image but then don't save the changes

Comment (andy-t-wang) [reply web/src/api/images/delete_all_images.ts:29]: Aligned in slack pushing this to another version

File: web/src/api/images/get_uploaded_image.ts
Comment (andy-t-wang) [reply web/src/api/images/get_uploaded_image.ts:59]: ok I am now using the session variables with the hasura actions is that secure?

File: web/src/api/images/get_unverified_images.ts
Comment (andy-t-wang) [reply web/src/api/images/get_unverified_images.ts:26]: Actually when you upload an image but don't save yet this won't be able to pull it. We also don't know what extension to manually get. Feel like at that point it might be easier to separate the logic. Would require us to then have get_all_unverified generate signed URLs even if an image doesn't exist there.

File: web/next.config.js
Comment (andy-t-wang) [inline web/next.config.js:8]: @pdtfh It is unclear to me if we need to have region name in prod, but on my local dev I did need it.

File: web/src/api/images/delete_all_images.ts
Comment (andy-t-wang) [inline web/src/api/images/delete_all_images.ts:21]: Think this makes sense to implement the actual trigger in a seperate pr from the image upload process. Just leaving the endpoint here for now.

**Comment by andy-t-wang** [comment]
Note: Since your last review I updated to hasura custom actions and added a img validation trigger back to unverified apps

File: hasura/migrations/default/1705775389481_turn_on_img_trigger/up.sql
Comment (pdtfh) [inline hasura/migrations/default/1705775389481_turn_on_img_trigger/up.sql:41]: likely also want to drop validate_all_img_urls_format?

File: web/next.config.js
Comment (andy-t-wang) [reply web/next.config.js:21]: resolved with security

File: web/src/api/images/delete_all_images.ts
Comment (pdtfh) [inline web/src/api/images/delete_all_images.ts:135]:
```suggestion
                  Key: "expired",
```

File: web/src/api/images/delete_all_images.ts
Comment (pdtfh) [inline web/src/api/images/delete_all_images.ts:136]:
```suggestion
                  Value: "true",
```

File: web/src/api/images/get_unverified_images.ts
Comment (pdtfh) [inline web/src/api/images/get_unverified_images.ts:48]: ok admin

File: web/src/api/images/get_unverified_images.ts
Comment (pdtfh) [inline web/src/api/images/get_unverified_images.ts:97]: In general is best to return more generic errors for permissions, e.g. "App not found"

File: web/src/api/images/get_uploaded_image.ts
Comment (pdtfh) [inline web/src/api/images/get_uploaded_image.ts:49]: same

File: web/src/api/images/get_uploaded_image.ts
Comment (pdtfh) [inline web/src/api/images/get_uploaded_image.ts:18]: wouldn't we want to get this automatically from Hasura? seems like it would not be great UX otherwise

File: web/src/api/images/upload_image_url.ts
Comment (andy-t-wang) [reply web/src/api/images/upload_image_url.ts:117]: yes it is the current interaction is using jpeg from the front end. Will update to have type checking

File: hasura/metadata/actions.yaml
Comment (andy-t-wang) [reply hasura/metadata/actions.yaml:24]: Was the one automatically being used when I made the action in the console

File: hasura/metadata/actions.yaml
Comment (andy-t-wang) [reply hasura/metadata/actions.yaml:50]: does not appear to be an option

File: hasura/metadata/actions.yaml
Comment (andy-t-wang) [reply hasura/metadata/actions.yaml:6]: Yeah originally didn't include it since it's not implemented, but added user and api_key

**Comment by andy-t-wang** [comment]
Can't find the comment anymore but within next.config we do need to use the hostname domain part not the full url. The image won't load if we do 
```
 {
        protocol: "https",
        hostname: "s3.us-east-2.amazonaws.com/developer-portal-2.0",
      },
```

File: web/next.config.js
Comment (andy-t-wang) [reply web/next.config.js]: This does not work with the presigned url domain returned by presignedpost

File: web/src/api/images/graphql/checkUserInApp.graphql
Comment (andy-t-wang) [reply web/src/api/images/graphql/checkUserInApp.graphql:1]: combined into a single query

File: web/src/api/images/upload_image_url.ts
Comment (andy-t-wang) [inline web/src/api/images/upload_image_url.ts:74]: We enforce jpeg on the front end by extracting form the file.type

**Comment by andy-t-wang** [comment]
Notes: Before Merging
1. Will update next.config to use a path name
2. Remove web-app-router .env

File: web/src/api/images/upload_image_url.ts
Comment (pdtfh) [inline web/src/api/images/upload_image_url.ts:104]:
```suggestion
        code: "not_found",
```

File: web/src/api/images/delete_all_images.ts
Comment (pdtfh) [inline web/src/api/images/delete_all_images.ts:93]: shouldn't admin also be able to do this?

Comment (pdtfh) [inline web/src/api/images/delete_all_images.ts:93]: another minor improvement, you could already account for roles in the query

**Review by pdtfh — APPROVED** [review]

File: web/src/api/images/delete_all_images.ts
Comment (andy-t-wang) [reply web/src/api/images/delete_all_images.ts:93]: This deletes all images and should only be used if the app is being deleted, so admin users. We will make a separate endpoint to delete individual images.

Comment (andy-t-wang) [reply web/src/api/images/delete_all_images.ts:93]: This query is used by getunverified images too. I added a roles check to the other query though

**Pull request merged**
